### ISSUE TITLE
pushNotes fails when builds run in parallel

### ISSUE THREAD

**Issue by skeggse** [semantic-release/semantic-release#1613]
## Current behavior

If two copies of `semantic-release` run concurrently, they can [trample each other's release notes](https://github.com/bee-queue/arena/runs/952013202?check_suite_focus=true) (stack trace excerpted below). I wonder if there's any way to configure a unique set of refs based on some CI-dependent metadata to avoid conflicts like these. This seems to have occurred because I had the primary branch and a maintenance branch attempt to release at the same time.

<details>
<summary>Stack trace</summary>

```
2020-08-06T02:42:23.3816391Z [2:42:23 AM] [semantic-release] › ✖  An error occurred while running semantic-release: Error: Command failed with exit code 1: git push https://x-access-token:[secure]@github.com/bee-queue/arena.git refs/notes/semantic-release
2020-08-06T02:42:23.3817588Z To https://github.com/bee-queue/arena.git
2020-08-06T02:42:23.3818767Z  ! [rejected]        refs/notes/semantic-release -> refs/notes/semantic-release (fetch first)
2020-08-06T02:42:23.3819565Z error: failed to push some refs to 'https://github.com/bee-queue/arena.git'
2020-08-06T02:42:23.3820132Z hint: Updates were rejected because the remote contains work that you do
2020-08-06T02:42:23.3820628Z hint: not have locally. This is usually caused by another repository pushing
2020-08-06T02:42:23.3821381Z hint: to the same ref. You may want to first integrate the remote changes
2020-08-06T02:42:23.3822140Z hint: (e.g., 'git pull ...') before pushing again.
2020-08-06T02:42:23.3822885Z hint: See the 'Note about fast-forwards' in 'git push --help' for details.
2020-08-06T02:42:23.3823762Z     at makeError (/home/runner/work/arena/arena/node_modules/execa/lib/error.js:59:11)
2020-08-06T02:42:23.3824236Z     at handlePromise (/home/runner/work/arena/arena/node_modules/execa/index.js:114:26)
2020-08-06T02:42:23.3825032Z     at processTicksAndRejections (internal/process/task_queues.js:93:5)
2020-08-06T02:42:23.3825765Z     at async pushNotes (/home/runner/work/arena/arena/node_modules/semantic-release/lib/git.js:248:3)
2020-08-06T02:42:23.3826681Z     at async run (/home/runner/work/arena/arena/node_modules/semantic-release/index.js:194:5)
2020-08-06T02:42:23.3827442Z     at async module.exports (/home/runner/work/arena/arena/node_modules/semantic-release/index.js:260:22)
2020-08-06T02:42:23.3828211Z     at async module.exports (/home/runner/work/arena/arena/node_modules/semantic-release/cli.js:55:5) {
2020-08-06T02:42:23.3829096Z   shortMessage: 'Command failed with exit code 1: git push https://x-access-token:[secure]@github.com/bee-queue/arena.git refs/notes/semantic-release',
2020-08-06T02:42:23.3829917Z   command: 'git push https://x-access-token:[secure]@github.com/bee-queue/arena.git refs/notes/semantic-release',
2020-08-06T02:42:23.3830363Z   exitCode: 1,
2020-08-06T02:42:23.3830725Z   signal: undefined,
2020-08-06T02:42:23.3831086Z   signalDescription: undefined,
2020-08-06T02:42:23.3831611Z   stdout: '',
2020-08-06T02:42:23.3832277Z   stderr: 'To https://github.com/bee-queue/arena.git\n' +
2020-08-06T02:42:23.3833027Z     ' ! [rejected]        refs/notes/semantic-release -> refs/notes/semantic-release (fetch first)\n' +
2020-08-06T02:42:23.3834290Z     "error: failed to push some refs to 'https://github.com/bee-queue/arena.git'\n" +
2020-08-06T02:42:23.3835034Z     'hint: Updates were rejected because the remote contains work that you do\n' +
2020-08-06T02:42:23.3835975Z     'hint: not have locally. This is usually caused by another repository pushing\n' +
2020-08-06T02:42:23.3836890Z     'hint: to the same ref. You may want to first integrate the remote changes\n' +
2020-08-06T02:42:23.3837566Z     "hint: (e.g., 'git pull ...') before pushing again.\n" +
2020-08-06T02:42:23.3838266Z     "hint: See the 'Note about fast-forwards' in 'git push --help' for details.",
2020-08-06T02:42:23.3838686Z   failed: true,
2020-08-06T02:42:23.3839662Z   timedOut: false,
2020-08-06T02:42:23.3840032Z   isCanceled: false,
2020-08-06T02:42:23.3840303Z   killed: false
2020-08-06T02:42:23.3840720Z }
```
</details>

## Expected behavior

No failure, no concurrent use of the same resource.

## Environment

- **semantic-release** version: 17.1.1
- CI environment: GitHub actions
- Plugins used: [several](https://github.com/bee-queue/arena/blob/07393fa293882d0ca90008fca4e00310a547526b/package.json#L21-L27), but the stack trace originates in `semantic-release` core
- **semantic-release** configuration: [release.config.js](https://github.com/bee-queue/arena/blob/master/release.config.js)
- [CI logs](https://github.com/bee-queue/arena/runs/952014075?check_suite_focus=true)

**Comment by varl** [comment]
In our organization we have seen this problem as well. Is there a workaround?

**Comment by ChristianUlbrich** [comment]
Apparently the way semantic-release is currently built, there is currently no easy way. The main culprit is, that there is a **single** _hard-coded_ (`refs/notes/semantic-release`) _git notes_ branch for **all** releases and furthermore a dependency on git notes itself. There will always be a problem, if you want to commit onto the same ref, which happens in exactly the situation that you describe (_simultaneous_ releases).

Without digging further into the code, the only solutions I see, is **either** making the usage of git notes _optional_ (and thus the whole release note logic) **or** using separate notes for each release branch. I think the latter is the easiest option™, but again without further digging into the code neither. :)

**Comment by varl** [comment]
Thanks for looking into it @ChristianUlbrich. 

I can add my (embarrassingly crude) workaround:

- Run the failing release job again.

**Comment by YitzyD** [comment]
Any update on whether the usage of the git notes feature configurable. Being able to disable git notes would be helpful for monorepos.

**Comment by fdmota** [comment]
I have the same issue

**Comment by travi** [comment]
closing as resolved by v23.0.0. please let us know if you find that to not be the case
